File: Sample Algorithms/beaupeop.py
def index(dp,n,u):
	for i in range(u-1,-1,-1):
		if(dp[i]==n):
			return i
	return -1
def seq(a,prev,i,s):
	if(prev[i]==-1):
		s.append(a[i])
		return s
	else:
		s.append(a[i])
		seq(a,prev,prev[i],s)
def lis(a,n):
	dp=[1]*n
	prev=[-1]*n
	for i in range(1,n):
		for j in range(i):
			if(a[i][1]>a[j][1] and a[i][0]>a[j][0] and dp[i]<dp[j]+1):
				dp[i]=dp[j]+1
				prev[i]=j
	i=dp.index(max(dp))
	s=[]
	seq(a,prev,i,s)
	return [max(dp),s]

#An application of longest increasing subsequence
n=int(input())
a=[]
for z in range(n):
	a.append(list(map(int,input().split()))+[z+1])
#a.sort()
#Sorted on the basis of strength
#Find longest increasing subsequence on the basis of beauty
m,s=lis(a,n)
print(m)
for i in s[::-1]:
	print(i[2],end=" ")
print("")

File: Sample Algorithms/beaupeop.cpp
#include <bits/stdc++.h>
using namespace std;

void seq(vector < pair<pair<int,int> ,int> > a,int *prev,int i,vector < int > &s){
	s.push_back(a[i].second);
	if(prev[i]==-1){
		return;
	}
	seq(a,prev,prev[i],s);
		
}
int lis(vector < pair<pair<int,int> ,int> > a,int n, vector < int > &s){
	int *dp=new int[n];
	int *prev=new int[n];
	for(int i=0;i<n;i++){
		dp[i]=1;
		prev[i]=-1;
	}
	for(int i=1;i<n;i++){
		for(int j=0;j<i;j++){
			if(a[i].first.first>a[j].first.first && a[i].first.second>a[j].first.second && dp[i]<dp[j]+1){
				dp[i]=dp[j]+1;
				prev[i]=j;
			}
		}
	}
	int m=dp[0];
	int k=0;
	for(int i=1;i<n;i++){
		if(dp[i]>m){
			m=dp[i];
			k=i;
		}
	}
	seq(a,prev,k,s);
	return m;
}
int main(){
	int n;
	cin>>n;
	int h,t;
	vector < pair<pair<int,int> ,int> > a;
	for(int i=0;i<n;i++){
		cin>>h>>t;
		a.push_back(make_pair(make_pair(h,t),i+1));
	}
	sort(a.begin(),a.end());
	/*for(int i=0;i<n;i++){
		cout<<a[i].first.first<<","<<a[i].first.second<<","<<a[i].second<<endl;
	}*/
	vector < int > s;
	int k=lis(a,n,s);
	cout<<k<<endl;
	for(int i=s.size()-1;i>=0;i--){
		cout<<s[i]<<" ";
	}
	cout<<endl;
	return 0;
}
